sshd-core/src/main/java/org/apache/sshd/certificate/OpenSshCertificateBuilder.java

```diff
@@ -0,0 +1,273 @@
+        }
+
+        if (null == publicKey) {
+            throw new IllegalStateException("'publicKey' is required");
+        }
+
+    }
+
+    public OpenSshCertificate sign(KeyPair caKeypair) throws Exception {
+
+        validate();
+
+        final String publicKeyType = KeyUtils.getKeyType(publicKey);
+        final String certType = SIGNATURE_ALGORITHM_MAP.get(publicKeyType);
+
+        // only certain kind of keys can be OpenSSH Certificates
+        if (null == certType) {
+            throw new UnsupportedOperationException(
+                    "unsupported public key type '" + publicKeyType + "' for OpenSSH Certificate");
+        }
+
+        final OpenSshCertificateImpl cert = new OpenSshCertificateImpl();
+        cert.setKeyType(certType);
+        cert.setType(type.getType());
+        cert.setCertPubKey(publicKey);
+        cert.setSerial(serial);
+        cert.setId(id);
+        if (null != principals) {
+            cert.setPrincipals(new ArrayList<>(principals));
+        }
+        if (null != criticalOptions) {
+            cert.setCriticalOptions(new ArrayList<>(criticalOptions));
+        }
+        if (null != extensions) {
+            cert.setExtensions(new ArrayList<>(extensions));
+        }
+        cert.setValidAfter(validAfter);
+        cert.setValidBefore(validBefore);
+
+        cert.setCaPubKey(caKeypair.getPublic());
+
+        if (null != nonce) {
+            cert.setNonce(nonce);
+        } else {
+            final SecureRandom rand = new SecureRandom();
+            final byte[] tempNonce = new byte[32];
+            rand.nextBytes(tempNonce);
+            cert.setNonce(tempNonce);
+        }
+
+        final ByteArrayBuffer toBeSignedBuf = new ByteArrayBuffer();
+        toBeSignedBuf.putRawPublicKey(cert);
+
+        SignatureFactoriesManager signatureManager = new SignatureFactoriesManager() {
+
+            @Override
+            public void setSignatureFactories(List<NamedFactory<Signature>> factories) {
+                // don't need
+            }
+
+            @Override
+            @SuppressWarnings({ "unchecked", "rawtypes" }) // safe due to the hierarchy
+            public List<NamedFactory<Signature>> getSignatureFactories() {
+                return (List) NamedFactory.setUpBuiltinFactories(false, BaseBuilder.DEFAULT_SIGNATURE_PREFERENCE);
+            }
+
+        };
+
+        final KeyPairIdentity caIdentity = new KeyPairIdentity(signatureManager, null, caKeypair);
+
+        // create signature using CA Keypair and OpenSSH Certificate data
+        final Map.Entry<String, byte[]> sigEntry = caIdentity.sign(null, null, toBeSignedBuf.getCompactData());
+
+        // set the signature (need to prefix this data section with the sig algorithm used)
+        final ByteArrayBuffer tmpBuffer = new ByteArrayBuffer();
+        // signature algorithm
+        tmpBuffer.putString(sigEntry.getKey());
+        // signature
+        tmpBuffer.putBytes(sigEntry.getValue());
```

Comment:
This signing is a bit roundabout. Inlining the relevant part of KeyPairIdentity.sign() here would give something like
```
String algo = KeyUtils.getKeyType(caKeypair.getPublic());
SignatureFactory factory = BuiltinSignatures.fromFactoryName(algo);
Signature signer = factory == null ? null : factory.create();
ValidateUtils.checkNotNull(signer, "No signer could be located for key type=%s", algo);
signer.initSigner(null, caKeypair.getPrivate());
signer.update(null, toBeSignedBuf.getCompactData());

byte[] signature = signer.sign(null);

// set the signature (need to prefix this data section with the sig algorithm used)
final ByteArrayBuffer tmpBuffer = new ByteArrayBuffer();
// signature algorithm
tmpBuffer.putString(algo);
// signature
tmpBuffer.putBytes(signature);
```
Then this builder could be moved to sshd-common, too.